UI: Refactor to combine the simpler cached reducers

maxlang opened the pull request:
Doesn't include tests for the generic reducer currently, but keeps
the old tests which ran against each individual reducer.

NOTE: If this approach looks good, I can consolidate the testing.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7506)

<!-- Reviewable:end -->

Commit:
Refactor to combine the simpler cached reducers

tamird commented:
approach LGTM. let's prune these tests down

---

Reviewed 17 of 17 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/tsconfig.json, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLN5vdl_JYsePHnfuaV:-KLN5vdmi_gaSjEY8zlK:677713193) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/tsconfig.json#L13)):*
> ```JSON
>     "app/app.tsx"
>   ]
> }
> ```

huh?

---

*[ui/app/redux/cachedDataReducer.ts, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCZJsohMZGtviB6gI:-KLNCZJsohMZGtviB6gJ:-391304344) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L3)):*
> ```XML
> /**
>  * This module maintains the state of read-only data about the cluster.
>  * Data is fetched from the '/_admin/v1/cluster' endpoint via 'util/api'.
> ```

looks like copy pasta? also "cluster" is everywhere in this file, which should be a more generic name.

---

*[ui/app/redux/cachedDataReducer.ts, line 6 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCi_QUSfPZqgKP_Hp:-KLNCi_QUSfPZqgKP_Hq:311338566) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L6)):*
> ```XML
>  */
> 
> import "whatwg-fetch";
> ```

why is this here?

---

*[ui/app/redux/cachedDataReducer.ts, line 11 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCWWlLgZXwnM6oGS9:-KLNCWWlLgZXwnM6oGSA:190015926) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L11)):*
> ```XML
> import { Action, PayloadAction } from "../interfaces/action";
> 
> interface Function {
> ```

what's this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-:-KLNDURwjXafr91JCoef:-621485914)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Will do.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/tsconfig.json, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLN5vdl_JYsePHnfuaV:-KLNEIFdaRnjzOEyVx_F:677916264) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/tsconfig.json#L13)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> huh?
</details>
oops

---

*[ui/app/redux/cachedDataReducer.ts, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCZJsohMZGtviB6gI:-KLNDqFVry9T318AP2-3:1421316155) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L3)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> looks like copy pasta? also "cluster" is everywhere in this file, which should be a more generic name.
</details>
Whoops, forgot to delete the copy pasta.

---

*[ui/app/redux/cachedDataReducer.ts, line 6 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCi_QUSfPZqgKP_Hp:-KLNDvMrY9l08wGs1s22:404585776) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L6)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why is this here?
</details>
I think it's needed for Promise. Let me doublecheck.

---

*[ui/app/redux/cachedDataReducer.ts, line 11 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCWWlLgZXwnM6oGS9:-KLNDz5KZvvO1k-8aTFr:1652106571) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L11)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's this?
</details>
Whoops, unused. Was trying to automatically generate the action strings using the api call function name, but TS was giving me an error when I tried to access `.name` on a function. I decided to be explicit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-:-KLNEJL7-42ygXppRTyz:1085483994)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[ui/app/redux/cachedDataReducer.ts, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCZJsohMZGtviB6gI:-KLNY6Ag7t2mehnbGYmj:-1160261952) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L3)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Whoops, forgot to delete the copy pasta.
</details>
There is a real lack of comments here.  A re-usable reducer is fairly

---

*[ui/app/redux/cachedDataReducer.ts, line 6 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCi_QUSfPZqgKP_Hp:-KLNX0dv2SgIQ9NI0joD:1832171901) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L6)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> I think it's needed for Promise. Let me doublecheck.
</details>
This statement ensures that "fetch" is available in the compiled javascript, by forcing whatwg-fetch to be included.  That JS module makes fetch globally available.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[ui/app/redux/cachedDataReducer.ts, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCZJsohMZGtviB6gI:-KLN_ax269LytgT6jROy:-893129151) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L3)):*
> able reducer is fairly
is a fairly complicated concept, let's make sure it's understood.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[ui/app/redux/cachedDataReducer.ts, line 6 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCi_QUSfPZqgKP_Hp:-KLNc5DinytrFIVDUINR:508492056) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L6)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> This statement ensures that "fetch" is available in the compiled javascript, by forcing whatwg-fetch to be included.  That JS module makes fetch globally available.
</details>
Cool story but it's still not needed here - it should be transitively included via api.ts.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[ui/app/redux/cachedDataReducer.ts, line 6 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCi_QUSfPZqgKP_Hp:-KLNsobvtku_dn6C-Een:-2002094585) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L6)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Cool story but it's still not needed here - it should be transitively included via api.ts.
</details>
Sure thing bro

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Added a couple more reducers and wrapped everything in the `cachedData` attribute at the top level in the state.

---

Review status: 5 of 24 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/redux/cachedDataReducer.ts, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCZJsohMZGtviB6gI:-KLY8FJQN81DHzo-QAOw:841980330) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L3)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> > able reducer is fairly
> is a fairly complicated concept, let's make sure it's understood.
</details>
Added some comments. Let me know if there's anything else you'd like to add.

---

*[ui/app/redux/cachedDataReducer.ts, line 6 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCi_QUSfPZqgKP_Hp:-KLY8KursTUKnICHMrYo:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L6)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Sure thing bro
</details>
Done.

---

*[ui/app/redux/cachedDataReducer.ts, line 11 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCWWlLgZXwnM6oGS9:-KLY8N28STeFC6Qc3HOk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L11)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Whoops, unused. Was trying to automatically generate the action strings using the api call function name, but TS was giving me an error when I tried to access `.name` on a function. I decided to be explicit.
</details>
Done.

---

*[ui/tsconfig.json, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLN5vdl_JYsePHnfuaV:-KLY9Fa2KMSJicsVb4Q9:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/tsconfig.json#L13)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> oops
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-:-KLY9LRN84jcWdaKOHGb:1643286825)*
<!-- Sent from Reviewable.io -->

tamird commented:
Have you tested this locally?

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 20 of 20 files at r2.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[ui/app/containers/events.tsx, line 92 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYBRxjGoRbph2nI5N1:-KLYBRxjGoRbph2nI5N2:1284624392) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/containers/events.tsx#L92)):*
> ```XML
> }
> 
> let events = (state: any): Event[] => state.cachedData.events.data && state.events.data.events;
> ```

what's the difference between state.cachedData.events and state.events? looks like a typo to me. good ol `any`

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 14 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYFKzWdok3IpmGD6tA:-KLYFKzWdok3IpmGD6tB:262468467) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.spec.ts#L14)):*
> ```XML
> }
> 
> let apiEndpointMock = (req = new Request(null)) => new Promise((resolve, reject) => resolve(new Response(req.request)));
> ```

:thumbsup: 

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 64 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYFW3RZantU8IbCono:-KLYFW3RZantU8IbConp:12956225) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.spec.ts#L64)):*
> ```XML
> 
>     it("should correctly dispatch receiveData", function () {
>       let r = new Response("test string");
> ```

`r` is the first letter of both `request` and `response`. how about `expectedResponse`?

also, what's the point of the "test string"? seems like this could just be null. chekhov's gun comes to mind.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 71 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYFQHlrmGD4NfovWhc:-KLYFQHlrmGD4NfovWhd:2125992474) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.spec.ts#L71)):*
> ```XML
>         valid: true,
>         data: r,
>         lastError: <Error>null,
> ```

inconsistent application of type parameter, probably don't need it.

---

*[ui/app/redux/cachedDataReducer.ts, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCZJsohMZGtviB6gI:-KLYG4afoJZc8Ejq3b2T:1313931292) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L3)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Added some comments. Let me know if there's anything else you'd like to add.
</details>
Nice comments.

---

*[ui/app/redux/cachedDataReducer.ts, line 11 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCWWlLgZXwnM6oGS9:-KLYCg2pK0BSauBZdZeo:259153557) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L11)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Done.
</details>
looks like it's still here.

---

*[ui/app/redux/cachedDataReducer.ts, line 37 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYDcC9u4JA4PiV_qcH:-KLYDcC9u4JA4PiV_qcI:1913439825) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L37)):*
> ```XML
> export class CachedDataReducer<TRequest, TResponseMessage> {
>   // Track all the currently seen namespaces, to ensure there isn't a conflict
>   public static namespaces: { [namespace: string]: boolean } = {};
> ```

public?

---

*[ui/app/redux/cachedDataReducer.ts, line 52 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEGd_3xsKvv1V-SxG:-KLYEGd_3xsKvv1V-SxH:-754953950) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L52)):*
> ```XML
>   constructor(private apiEndpoint: (req: TRequest) => Promise<TResponseMessage>, public namespace: string, private invalidationPeriodMillis?: number) {
>     // check namespace
>     assert.notProperty(CachedDataReducer.namespaces, namespace, "Expect namespace to be unique.");
> ```

expected?

---

*[ui/app/redux/cachedDataReducer.ts, line 53 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYE8D1u70XO16B0Um4:-KLYE8D1u70XO16B0Um5:-2028884177) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L53)):*
> ```XML
>     // check namespace
>     assert.notProperty(CachedDataReducer.namespaces, namespace, "Expect namespace to be unique.");
>     CachedDataReducer.namespaces[namespace] = true;
> ```

`namespace` is a reserved word in TS - i guess the parser is smart enough here, but this still messes with syntax highlighting. can you pick a different name?

---

*[ui/app/redux/cachedDataReducer.ts, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEVRkUA4vcw4q70j8:-KLYEVRkUA4vcw4q70j9:1286670636) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L133)):*
> ```XML
>    * - a request is not in flight AND
>    *   - its results are not considered valid OR
>    *   - it has no invalidation period
> ```

i don't think i grok this; if there's no invalidation period then this thing will attempt to fetch fresh data anyway?

perhaps this is clarified by replace "as long as" with "iff" above.

---

*[ui/app/redux/cachedDataReducer.ts, line 139 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEirrHpDOwvMoruao:-KLYEirrHpDOwvMoruap:2070478551) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L139)):*
> ```XML
>       let state: CachedDataReducerState<TResponseMessage> = getState().cachedData[this.namespace];
> 
>       // Don't refresh if a request is already in flight
> ```

comment is outdated, but also useless. remove?

---

*[ui/app/redux/cachedDataReducer.ts, line 146 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEwZCNF9LWVV3yCta:-KLYEwZCNF9LWVV3yCtb:1394414062) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L146)):*
> The promise is returned for testing.

What does this mean?

---

*[ui/app/util/api.ts, line 142 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYCNWa1yFzjaV_oKz6:-KLYCNWa1yFzjaV_oKz7:-1955843013) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/util/api.ts#L142)):*
> ```XML
>   return Fetch(serverpb.NodesResponse, `/_status/nodes`).then((r: NodesResponseMessage) => {
>   return _.map(r.nodes, (node) => {
>       RollupStoreMetrics(node);
> ```

why break the convention? i'd rather see a method in util/proto that wraps this guy.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-:-KLYBi_GlZQdRAZpQK91:518803814)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Yeah I'm going through that now. Looks like there's an issue with the nodes reducer.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-:-KLYGYTRSifgF1sNfeMB:-1799937308)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 20 of 24 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/events.tsx, line 92 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYBRxjGoRbph2nI5N1:-KLYGn-flOuLQo9LSvIV:553059518) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/containers/events.tsx#L92)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's the difference between state.cachedData.events and state.events? looks like a typo to me. good ol `any`
</details>
I added a level of indirection. I now include all the apiReducers under a `cachedData` attribute in the state.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 64 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYFW3RZantU8IbCono:-KLYHM3wuIoBbj3e_OW2:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.spec.ts#L64)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `r` is the first letter of both `request` and `response`. how about `expectedResponse`?
> 
> also, what's the point of the "test string"? seems like this could just be null. chekhov's gun comes to mind.
</details>
Done.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 71 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYFQHlrmGD4NfovWhc:-KLYHiXoOpaTMAk2fRRw:-1267191076) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.spec.ts#L71)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> inconsistent application of type parameter, probably don't need it.
</details>
You do need it because otherwise lastError is of type `any`. I've specified the type of `expected` to avoid this. 

---

*[ui/app/redux/cachedDataReducer.ts, line 11 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLNCWWlLgZXwnM6oGS9:-KLYI2kwcDknlw6e5T3i:-157194139) ([raw file](https://github.com/cockroachdb/cockroach/blob/58b7220a96f706635a53695564c7aca9a2fd5df1/ui/app/redux/cachedDataReducer.ts#L11)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> looks like it's still here.
</details>
Whoops.... some sort of magic blindness prevented me from seeing this

---

*[ui/app/redux/cachedDataReducer.ts, line 37 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYDcC9u4JA4PiV_qcH:-KLYIi59BKyyocR1panV:-425215373) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L37)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> public?
</details>
Good point, private works fine.

---

*[ui/app/redux/cachedDataReducer.ts, line 52 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEGd_3xsKvv1V-SxG:-KLYKBSGNV1ABv7Wiy9L:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L52)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> expected?
</details>
Done.

---

*[ui/app/redux/cachedDataReducer.ts, line 53 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYE8D1u70XO16B0Um4:-KLYK4ItynTO7bAoqwnb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L53)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `namespace` is a reserved word in TS - i guess the parser is smart enough here, but this still messes with syntax highlighting. can you pick a different name?
</details>
Done.

---

*[ui/app/redux/cachedDataReducer.ts, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEVRkUA4vcw4q70j8:-KLYNSPQL-ChbZd1WdLJ:1675258936) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L133)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i don't think i grok this; if there's no invalidation period then this thing will attempt to fetch fresh data anyway?
> 
> perhaps this is clarified by replace "as long as" with "iff" above.
</details>
If there's no invalidation period then currently there's no way to invalidate the data, so we just refresh it every time.

Done.

---

*[ui/app/redux/cachedDataReducer.ts, line 139 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEirrHpDOwvMoruao:-KLYRsXuEktNE4Av8RDp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L139)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> comment is outdated, but also useless. remove?
</details>
Done.

---

*[ui/app/redux/cachedDataReducer.ts, line 146 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEwZCNF9LWVV3yCta:-KLYRxwwPAMyMAwZtSXL:1695295141) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L146)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> > The promise is returned for testing.
> 
> What does this mean?
</details>
The returned promise is only used in the tests. Rephrased a little.

---

*[ui/app/util/api.ts, line 142 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYCNWa1yFzjaV_oKz6:-KLYUniRG6PdBIA-0mUN:1768007267) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/util/api.ts#L142)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why break the convention? i'd rather see a method in util/proto that wraps this guy.
</details>
This API is never used raw. I considered it part of the preprocessing necessary to use this endpoint. I think this makes more sense as the convention. @mrtracy?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/events.tsx, line 92 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYBRxjGoRbph2nI5N1:-KLY_3Mzkk7wYahKzpso:-816609316) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/containers/events.tsx#L92)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> I added a level of indirection. I now include all the apiReducers under a `cachedData` attribute in the state.
</details>
Are you saying that the specific reducer is *also* present under `state` at the top level? Having two ways to do things seems like something we'd like to avoid, especially in cases like this one where we don't have the help of the type checker.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 71 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYFQHlrmGD4NfovWhc:-KLYZNKdNP5mbgcE0JWU:-340745835) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.spec.ts#L71)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> You do need it because otherwise lastError is of type `any`. I've specified the type of `expected` to avoid this. 
</details>
Oh. Can you specify the type of `expected` throughout? you could probably do that way at the top, where you specify the type of `state`

---

*[ui/app/redux/cachedDataReducer.ts, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEVRkUA4vcw4q70j8:-KLYZfRU1zljWoyHoS9D:36925055) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L133)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> If there's no invalidation period then currently there's no way to invalidate the data, so we just refresh it every time.
> 
> Done.
</details>
heh, sounds equivalent to an invalidation period of zero. sounds like you can remove nullability from that API?

---

*[ui/app/util/api.ts, line 142 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYCNWa1yFzjaV_oKz6:-KLYZz5axJ6JkxVpJTOZ:1722153678) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/util/api.ts#L142)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> This API is never used raw. I considered it part of the preprocessing necessary to use this endpoint. I think this makes more sense as the convention. @mrtracy?
</details>
Please no. Just give it a thin wrapper and keep these concerns separate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[ui/app/containers/events.tsx, line 92 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYBRxjGoRbph2nI5N1:-KLZ0TrhXij8l3tQTHQc:-542507145) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/containers/events.tsx#L92)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Are you saying that the specific reducer is *also* present under `state` at the top level? Having two ways to do things seems like something we'd like to avoid, especially in cases like this one where we don't have the help of the type checker.
</details>
This definitely looks incorrect.  `state` doesn't appear to have an `events` at the top level.

---

*[ui/app/util/api.ts, line 142 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYCNWa1yFzjaV_oKz6:-KLYzeRGYGvoWcj546tn:389499369) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/util/api.ts#L142)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Please no. Just give it a thin wrapper and keep these concerns separate.
</details>
It was appropriate to put this logic in this function when it was co-located with its reducer. However, the api.ts file is entirely concerned with pulling things off of the network; this is not a fix for a network problem, but rather a desirable convenience for pages that use this data.  It would be better in the reducer file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
LGTM with a few comments.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-:-KLZ0l8aIno7ZcL-n9Ou:829508448)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 16 of 24 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/events.tsx, line 92 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYBRxjGoRbph2nI5N1:-KLb-Z1KYx1EjZEmdUeM:1143967908) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/containers/events.tsx#L92)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> This definitely looks incorrect.  `state` doesn't appear to have an `events` at the top level.
</details>
I misunderstood.

Done.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 71 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYFQHlrmGD4NfovWhc:-KLb1KubltDObaR5pAFt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.spec.ts#L71)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Oh. Can you specify the type of `expected` throughout? you could probably do that way at the top, where you specify the type of `state`
</details>
Done.

---

*[ui/app/redux/cachedDataReducer.ts, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEVRkUA4vcw4q70j8:-KLYcepk0-4pz3Xz2XlI:1804175215) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L133)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> heh, sounds equivalent to an invalidation period of zero. sounds like you can remove nullability from that API?
</details>
The difference is that `0` means it would constantly be refreshed.

Invalidation period of `0`:
1) React Component is mounted and data is refreshed
2) Date is now valid 
3) Component receives new props and displays data
4) Data is instantly invalidated after 1 tick 
5) React Component receives props and dispatches refresh
6) Proceed to step 2

Invalidation period of `null`:
1) React Component is mounted and data is refreshed
2) Date is now valid 
3) Component receives new props and displays data

If you navigate away and back, the component is mounted again and the data is refreshed at that time.

---

*[ui/app/util/api.ts, line 142 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYCNWa1yFzjaV_oKz6:-KLYeY5NHITIo5Jr6slA:203522279) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/util/api.ts#L142)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> It was appropriate to put this logic in this function when it was co-located with its reducer. However, the api.ts file is entirely concerned with pulling things off of the network; this is not a fix for a network problem, but rather a desirable convenience for pages that use this data.  It would be better in the reducer file.
</details>
Moved to `apiReducers.ts`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 71 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYFQHlrmGD4NfovWhc:-KLb4ApZdn5AWcxqA-sX:1718694397) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.spec.ts#L71)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Done.
</details>
does this syntax still work? why do you need
```
expected = new ...
expected.foo = bar
```
elsewhere? can you stick to one or the other throughout?

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 49 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLb3zHGtuBipmniyE7X:-KLb3zHGtuBipmniyE7Y:302946204) ([raw file](https://github.com/cockroachdb/cockroach/blob/736c4fb6883308ad1b2ff3ab0fab82f00a2df3b4/ui/app/redux/cachedDataReducer.spec.ts#L49)):*
> ```XML
> 
>     it("should have the correct default value.", function() {
>       expected = new CachedDataReducerState<Response>();
> ```

nit: can't this be `new CachedDataReducerState()` throughout?

---

*[ui/app/redux/cachedDataReducer.ts, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEVRkUA4vcw4q70j8:-KLb4_mrLlZlNmeC9TfN:-284351022) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L133)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> The difference is that `0` means it would constantly be refreshed.
> 
> Invalidation period of `0`:
> 1) React Component is mounted and data is refreshed
> 2) Date is now valid 
> 3) Component receives new props and displays data
> 4) Data is instantly invalidated after 1 tick 
> 5) React Component receives props and dispatches refresh
> 6) Proceed to step 2
> 
> Invalidation period of `null`:
> 1) React Component is mounted and data is refreshed
> 2) Date is now valid 
> 3) Component receives new props and displays data
> 
> If you navigate away and back, the component is mounted again and the data is refreshed at that time.
> 
</details>
:+1:, thanks.

---

*[ui/app/util/api.ts, line 7 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLb4LgKuaMFYE3UQLNP:-KLb4LgLi99bM0mMjksT:-1218784804) ([raw file](https://github.com/cockroachdb/cockroach/blob/02a236fd37de6caf0d9b9abb84d726d93d6f9e8e/ui/app/util/api.ts#L7)):*
> ```XML
> import * as _ from "lodash";
> import "whatwg-fetch";
> ```

does this file need any changes? seems like it can be left unchanged.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 22 of 23 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 71 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYFQHlrmGD4NfovWhc:-KLc2ZbGP_uuvqNdZPO3:1239983464) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.spec.ts#L71)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> does this syntax still work? why do you need
> ```
> expected = new ...
> expected.foo = bar
> ```
> elsewhere? can you stick to one or the other throughout?
</details>
Both syntaxes work. I went with what I felt was the most terse in each case, but I'll make it consistent.

The reason I couldn't leave the others with the old syntax was something that surprised me. It will complain if they don't have the `data` and `lastError` properties, but if they're included as `undefined` that isn't equivalent. The generated code for the `CachedDataReducerState` class is

```
var CachedDataReducerState = (function() {
  function CachedDataReducerState() {
    this.inFlight = false;
    this.valid = false;
  }
  return CachedDataReducerState;
}());
```

but typescript would complain if you actually tried to assign this to something of type `CachedDataReducerState`.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 49 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLb3zHGtuBipmniyE7X:-KLc2QtD3iL2Gsgh9uN-:-1362213331) ([raw file](https://github.com/cockroachdb/cockroach/blob/736c4fb6883308ad1b2ff3ab0fab82f00a2df3b4/ui/app/redux/cachedDataReducer.spec.ts#L49)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: can't this be `new CachedDataReducerState()` throughout?
</details>
I get `Type 'CachedDataReducerState<{}>' is not assignable to type 'CachedDataReducerState<Response>` when I try this.

---

*[ui/app/redux/cachedDataReducer.ts, line 133 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLYEVRkUA4vcw4q70j8:-KLc46yzg1GDQ4kcNeYI:998628876) ([raw file](https://github.com/cockroachdb/cockroach/blob/702cd708848ec8e5e75b459b5101b8f7eac94c3e/ui/app/redux/cachedDataReducer.ts#L133)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> :+1:, thanks.
</details>
s/Date/Data/g in my description, but shouldn't change anything.

---

*[ui/app/util/api.ts, line 7 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLb4LgKuaMFYE3UQLNP:-KLc4KIhf7wzDRrw3p37:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/02a236fd37de6caf0d9b9abb84d726d93d6f9e8e/ui/app/util/api.ts#L7)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> does this file need any changes? seems like it can be left unchanged.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> :shipit: 

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 49 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLb3zHGtuBipmniyE7X:-KLcPsbsmfqSSbYzwiDY:-665481080) ([raw file](https://github.com/cockroachdb/cockroach/blob/736c4fb6883308ad1b2ff3ab0fab82f00a2df3b4/ui/app/redux/cachedDataReducer.spec.ts#L49)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> I get `Type 'CachedDataReducerState<{}>' is not assignable to type 'CachedDataReducerState<Response>` when I try this.
</details>
Surprising. Thanks.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 108 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLcSFaKVAlvf2Oi2P2x:-KLcSFaKVAlvf2Oi2P2y:-1210263559) ([raw file](https://github.com/cockroachdb/cockroach/blob/e569451f70dbce9a9eda292164a91ec7bd30f544/ui/app/redux/cachedDataReducer.spec.ts#L108)):*
> ```XML
> 
>       return testReducerObj.refresh(new Request(testString))(dispatch, () => state).then(() => {
>         assert.deepEqual(state.cachedData.test, {
> ```

did you mean to use the `new ....` syntax here?

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 115 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLcSE_phT5UAcbxiKOO:-KLcSE_phT5UAcbxiKOP:-1280936503) ([raw file](https://github.com/cockroachdb/cockroach/blob/e569451f70dbce9a9eda292164a91ec7bd30f544/ui/app/redux/cachedDataReducer.spec.ts#L115)):*
> ```XML
>         });
>       });
> ```

remove

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-:-KLcSL4zOI9wK7aWPlQz:-558175116)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 19 of 23 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 108 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLcSFaKVAlvf2Oi2P2x:-KLcVbDNK0-2EdRYX603:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e569451f70dbce9a9eda292164a91ec7bd30f544/ui/app/redux/cachedDataReducer.spec.ts#L108)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> did you mean to use the `new ....` syntax here?
</details>
Done.

---

*[ui/app/redux/cachedDataReducer.spec.ts, line 115 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLcSE_phT5UAcbxiKOO:-KLcVsEgWFiqXTpze0k7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e569451f70dbce9a9eda292164a91ec7bd30f544/ui/app/redux/cachedDataReducer.spec.ts#L115)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> remove
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 20 of 24 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[ui/app/redux/apiReducers.ts, line 32 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506#-KLvL5xGyWP2p-2xmFcV:-KMGFNC1yvVBxeKjw5io:-341944961) ([raw file](https://github.com/cockroachdb/cockroach/blob/b52fd31d6bca9e4631d2da55aceb4855b04365ee/ui/app/redux/apiReducers.ts#L32)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how does this work? underscore is not imported in this file.
</details>
Done.

I'm not sure how it worked. I set a breakpoint at the `import` and `_` was `undefined`, but `global._` was already defined as `lodash`. Anyway, it's corrected now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7506)*
<!-- Sent from Reviewable.io -->
